pkg/phlaredb/block/fetcher.go

```diff
@@ -0,0 +1,606 @@
+	}, nil
+}
+
+var (
+	ErrorSyncMetaNotFound  = errors.New("meta.json not found")
+	ErrorSyncMetaCorrupted = errors.New("meta.json corrupted")
+)
+
+// loadMeta returns metadata from object storage or error.
+// It returns ErrorSyncMetaNotFound and ErrorSyncMetaCorrupted sentinel errors in those cases.
+func (f *MetaFetcher) loadMeta(ctx context.Context, id ulid.ULID) (*Meta, error) {
+	var (
+		metaFile       = path.Join(id.String(), MetaFilename)
+		cachedBlockDir = filepath.Join(f.cacheDir, id.String())
+	)
+
+	// Block meta.json file is immutable, so we lookup the cache as first thing without issuing
+	// any API call to the object storage. This significantly reduce the pressure on the object
+	// storage.
+	//
+	// Details of all possible cases:
+	//
+	// - The block upload is in progress: the meta.json file is guaranteed to be uploaded at last.
+	//   When we'll try to read it from object storage (later on), it will fail with ErrorSyncMetaNotFound
+	//   which is correctly handled by the caller (partial block).
+	//
+	// - The block upload is completed: this is the normal case. meta.json file still exists in the
+	//   object storage and it's expected to match the locally cached one (because it's immutable by design).
+	// - The block has been marked for deletion: the deletion hasn't started yet, so the full block (including
+	//   the meta.json file) is still in the object storage. This case is not different than the previous one.
+	//
+	// - The block deletion is in progress: loadMeta() function may return the cached meta.json while it should
+	//   return ErrorSyncMetaNotFound. This is a race condition that could happen even if we check the meta.json
+	//   file in the storage, because the deletion could start right after we check it but before the MetaFetcher
+	//   completes its sync.
+	//
+	// - The block has been deleted: the loadMeta() function will not be called at all, because the block
+	//   was not discovered while iterating the bucket since all its files were already deleted.
+	if m, seen := f.cached[id]; seen {
+		return m, nil
+	}
+
+	// Best effort load from local dir.
+	if f.cacheDir != "" {
+		m, err := ReadMetaFromDir(cachedBlockDir)
+		if err == nil {
+			return m, nil
+		}
+
+		if !errors.Is(err, os.ErrNotExist) {
+			level.Warn(f.logger).Log("msg", "best effort read of the local meta.json failed; removing cached block dir", "dir", cachedBlockDir, "err", err)
+			if err := os.RemoveAll(cachedBlockDir); err != nil {
+				level.Warn(f.logger).Log("msg", "best effort remove of cached dir failed; ignoring", "dir", cachedBlockDir, "err", err)
+			}
+		}
+	}
+
+	// todo(cyriltovena): we should use ReaderWithExpectedErrs(f.bkt.IsObjNotFoundErr) here, to avoid counting IsObjNotFoundErr as an error
+	// since this is expected
+	r, err := f.bkt.Get(ctx, metaFile)
+	if f.bkt.IsObjNotFoundErr(err) {
+		// Meta.json was deleted between bkt.Exists and here.
+		return nil, errors.Wrapf(ErrorSyncMetaNotFound, "%v", err)
+	}
```

Comment:
I guess it should be fine. The only caller is `fetchMetadata` which does handle `ErrorSyncMetaNotFound` error properly so we probably don't have to do anything special here

Reply:
Yeah the original code was doing this, mostly for metrics, which is not urgent, but I think I can roll it back I was able to make ReaderWithExpectedErrs work I think